core/src/main/java/org/apache/iceberg/encryption/StandardEncryptionManager.java

```
@@ -0,0 +1,97 @@
+   * @param kmsClient Client of KMS used to wrap/unwrap keys in envelope encryption
+   * @param encryptionProperties encryption properties
+   */
+  public StandardEncryptionManager(
+      String tableKeyId,
+      int dataKeyLength,
+      KeyManagementClient kmsClient,
+      Map<String, String> encryptionProperties) {
+    Preconditions.checkNotNull(tableKeyId, "Invalid encryption key ID: null");
+    Preconditions.checkNotNull(kmsClient, "Invalid KMS client: null");
+    Preconditions.checkNotNull(encryptionProperties, "Invalid encryption properties: null");
+    this.tableKeyId = tableKeyId;
+    this.kmsClient = kmsClient;
+
+    this.dataKeyLength = dataKeyLength;
+  }
+
+  @Override
+  public EncryptedOutputFile encrypt(OutputFile rawOutput) {
+    lazyCreateRNG();
+
+    ByteBuffer fileDek = ByteBuffer.allocate(dataKeyLength);
+    workerRNG.nextBytes(fileDek.array());
+
+    ByteBuffer aadPrefix = ByteBuffer.allocate(EncryptionProperties.ENCRYPTION_AAD_LENGTH_DEFAULT);
+    workerRNG.nextBytes(aadPrefix.array());
+
+    KeyMetadata encryptionMetadata = new KeyMetadata(fileDek, aadPrefix);
+
+    return new BaseEncryptedOutputFile(
+        new AesGcmOutputFile(rawOutput, fileDek.array(), aadPrefix.array()),
+        encryptionMetadata,
+        rawOutput);
+  }
+
+  @Override
+  public InputFile decrypt(EncryptedInputFile encrypted) {
+    Preconditions.checkNotNull(encrypted.keyMetadata().buffer(), "Invalid key metadata: null");
+
+    KeyMetadata keyMetadata = KeyMetadata.parse(encrypted.keyMetadata().buffer());
+
+    byte[] fileDek = keyMetadata.encryptionKey().array();
+    byte[] aadPrefix = keyMetadata.aadPrefix().array();
+
+    return new AesGcmInputFile(encrypted.encryptedInputFile(), fileDek, aadPrefix);
+  }
+
+  private void lazyCreateRNG() {
+    if (this.workerRNG == null) {
+      this.workerRNG = new SecureRandom();
+    }
+  }
+
+  public ByteBuffer wrapKey(ByteBuffer secretKey) {
+    return kmsClient.wrapKey(secretKey, tableKeyId);
```

These should throw an exception if `kmsClient` is `null` so that we ensure it is never called from workers.